github/client_repository_file.go

```diff
@@ -61,17 +74,26 @@ func (c *FileClient) Get(ctx context.Context, path, branch string, optFns ...git
 	for _, file := range directoryContent {
 		filePath := file.Path
 		if *file.Type == "dir" {
-			if recursive == true {
+			if fileOpts.Recursive == true {
+				// stop recursive calls when level is the max level reached
+				if fileOpts.MaxDepth == 1 {
+					break
+				}
+
 				if !strings.HasSuffix(path, "/") {
 					path = path + "/"
 				}
-				subdirectoryPath := fmt.Sprintf("%v%v/", path, *file.Name)
+				subdirectoryPath := fmt.Sprintf("%s%s/", path, *file.Name)
+
 				// recursive call for child directories to get their content
-				childFiles, err := c.Get(ctx, subdirectoryPath, branch, optFns...)
+				childOptFns := gitprovider.FilesGetOptions{Recursive: fileOpts.Recursive, MaxFiles: fileOpts.MaxFiles, MaxDepth: fileOpts.MaxDepth - 1}
+				childFiles, err := c.Get(ctx, subdirectoryPath, branch, &childOptFns)
+
 				if err != nil {
 					return nil, err
 				}
 				files = append(files, childFiles...)
+				files = limitSliceSize(files, fileOpts.MaxFiles)
```

Comment:
We want to limit the number of files but also the number of api calls. So if we have reached the maxFiles limit, we should break out of the loop and return the files.

`MaxFiles` feels like a hard limit, so we need to make sure we do not download more than that number. If it is a best effort, we should say so.